nameko/dependencies.py

```diff
@@ -68,75 +69,71 @@ def kill(self):
-        instance is created and then bound to the ServiceContainer instance
-        controlling its lifecyle.
+        """ Bind a clone of this extension to `container` with the given
+        `name`.
         """
-        self.name = name
-        self.container = container
-        self.bound = True
+        if self.__clone:  # TODO: prefer this over check in clone?
+            raise RuntimeError('Cloned extensions cannot be bound.')
+
+        # if this extension is shared and there's already a cloned instance,
+        # return that
+        if self.__shared:
+            shared_extensions.setdefault(container, {})
+            shared = shared_extensions[container].get(self.sharing_key)
+            if shared:
+                return shared
+
+        # clone this extension and associate it with the container
+        clone = self.clone()
+        clone.name = name
+        clone.container = container
+
+        # recursively bind nested extensions
+        for ext_name, ext in inspect.getmembers(self, is_extension):
+            setattr(clone, ext_name, ext.bind(ext_name, container))
+
+        # if this extension is shared, save the cloned instance
+        if self.__shared:
+            shared_extensions[container][self.sharing_key] = clone
+
+        return clone
+
+    def clone(self):
+        if self.__state is None:
+            raise RuntimeError('Undefined extension state. Did you forget to '
+                               'call super().__init__() in your subclass?')
+
+        if self.__clone:  # TODO: need this even with the check in bind above?
+            raise RuntimeError('Cloned extensions cannot be cloned.')
+
+        cls = type(self)
+        args, kwargs = self.__state
+        instance = cls(*args, **kwargs)
+        instance.__clone = True
+        # alternative: del clone.clone; del clone.bind
+        return instance
 
     @property
-    def nested_dependencies(self):
-        """ Recusively yield nested dependencies of DependencyProvider.
-
-        For example, with an instance of `:class:~nameko.rpc.RpcProvider`
-        called ``rpc``::
-
-            >>> deps = list(rpc.nested_dependencies)
-            >>> deps
-            [<nameko.rpc.RpcConsumer object at 0x10d125690>,
-             <nameko.messaging.QueueConsumer object at 0x10d5b8f50>]
-            >>>
+    def extensions(self):
+        """ A generator that yields all nested extensions, including `self`
         """
-        for _, attr in inspect.getmembers(self):
-            if isinstance(attr, DependencyProvider):
-                yield attr
-                for nested_dep in attr.nested_dependencies:
-                    yield nested_dep
+        for _, ext in iter_extensions(self):
+            yield ext
+        yield self
+
+    @property
+    def sharing_key(self):
+        return type(self)
 
     def __repr__(self):
-        if not self.bound:
-            return '<{} [unbound] at 0x{:x}>'.format(
+        if self.name is UNBOUND_NAME:
```

Comment:
a little odd to use a string as a sentinel. (i.e. should this just be `==` do we want stronger guarantees)